File: Euclid/Network/Session/ConnectionSession.cs
using System;
using System.Net;
using DotNetty.Transport.Channels;
using Euclid.Game;
using Euclid.Messages;
using Euclid.Util.Encryption;

namespace Euclid.Network.Session
{
    public class ConnectionSession
    {
        #region Fields

        private bool m_Disconnected;

        #endregion

        #region Properties

        /// <summary>
        /// Gets the player channel.
        /// </summary>
        public IChannel Channel { get; private set; }

        /// <summary>
        /// Get the ip address of the player connected.
        /// </summary>
        public string IpAddress { get; set; }

        /// <summary>
        /// Get player instance
        /// </summary>
        public Player Player { get; private set; }

        /// <summary>
        /// Encryption instance
        /// </summary>
        public RC4 Encryption { get; private set; }
        
        /// <summary>
        /// Public key for client-side deciphering
        /// </summary>
        public string PublicKey { get; private set; }
        
        /// <summary>
        /// Locally connected port
        /// </summary>
        public int LocalPort { get; private set; }

        /// <summary>
        /// Connection state on whether it is on MAIN, PRIVATE, PUBLIC
        /// </summary>
        public ConnectionMode Mode { get; private set; }

        #endregion

        #region Constructors

        /// <summary>
        /// Constructor for player.
        /// </summary>
        /// <param name="channel">the channel</param>
        public ConnectionSession(IChannel channel)
        {
            Channel = channel;
            IpAddress = ((IPEndPoint)channel.LocalAddress).Address.ToString();
            LocalPort = ((IPEndPoint)channel.LocalAddress).Port;
            PublicKey = Guid.NewGuid().ToString().Replace("-", "").ToLower();

            if (LocalPort == GameServer.Instance.MainServer.Port)
                Mode = ConnectionMode.MAIN;

            if (LocalPort == GameServer.Instance.PrivateServer.Port)
                Mode = ConnectionMode.PRIVATE;

            if (GameServer.Instance.PublicServers.Exists(x => x.Port == LocalPort))
                Mode = ConnectionMode.PUBLIC;

            Player = new Player(this);
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Send message composer
        /// </summary>
        public void Send(IMessageComposer composer)
        {
            if (!composer.Composed)
            {
                composer.Composed = true;
                composer.Write();
            }

            try
            {
                Channel.WriteAndFlushAsync(composer);
            }
            catch { }
        }

        public void InitialiseEncryption()
        {
            if (Encryption != null)
            {
                return;
            }

            Encryption = new RC4();
            Encryption.SetKey(SecretKey.SecretDecode(PublicKey));

            if (Euclid.Encryption)
            {
                // If the connection mode is the main server, or there is no main server connection then we add it
                // if (Mode == ConnectionMode.MAIN || Player.MainServer == null)
                Channel.Pipeline.AddFirst("encryption", new EncryptionDecoder(this.Encryption));
            }
        }

        /// <summary>
        /// Disconnection handler
        /// </summary>
        public virtual void OnDisconnect()
        {
            if (m_Disconnected)
                return;

            m_Disconnected = true;
            Player.OnDisconnect();
        }

        /// <summary>
        /// Forcibly disconnect player
        /// </summary>
        public virtual void Disconnect()
        {
            if (m_Disconnected)
                return;

            this.Channel.CloseAsync();
        }

        #endregion

    }
}

File: euclid/src/main/java/euclid/network/session/ConnectionSession.java
package euclid.network.session;

import euclid.Euclid;
import euclid.game.Player;
import euclid.messages.IMessageComposer;
import euclid.network.GameServer;
import euclid.network.streams.GameAddress;
import euclid.util.encryption.RC4;
import euclid.util.encryption.SecretKey;
import io.netty.channel.Channel;

import java.net.InetSocketAddress;
import java.util.List;
import java.util.UUID;

public class ConnectionSession {

    private boolean disconnected;

    private final Channel channel;
    private final String ipAddress;
    private final int localPort;
    private final String publicKey;
    private final Player player;
    private final ConnectionMode mode;

    private RC4 encryption;

    public ConnectionSession(Channel channel) {
        this.channel = channel;
        InetSocketAddress remoteAddress = (InetSocketAddress) channel.remoteAddress();
        this.ipAddress = remoteAddress != null ? remoteAddress.getAddress().getHostAddress() : "unknown";
        this.localPort = ((InetSocketAddress) channel.localAddress()).getPort();
        this.publicKey = UUID.randomUUID().toString().replace("-", "").toLowerCase();

        GameServer server = GameServer.getInstance();
        ConnectionMode resolvedMode = ConnectionMode.MAIN;

        if (server != null) {
            if (localPort == server.getMainServer().getPort()) {
                resolvedMode = ConnectionMode.MAIN;
            } else if (localPort == server.getPrivateServer().getPort()) {
                resolvedMode = ConnectionMode.PRIVATE;
            } else {
                List<GameAddress> publicServers = server.getPublicServers();
                for (GameAddress addr : publicServers) {
                    if (addr.getPort() == localPort) {
                        resolvedMode = ConnectionMode.PUBLIC;
                        break;
                    }
                }
            }
        }

        this.mode = resolvedMode;
        this.player = new Player(this);
    }

    public Channel getChannel() {
        return channel;
    }

    public String getIpAddress() {
        return ipAddress;
    }

    public Player getPlayer() {
        return player;
    }

    public RC4 getEncryption() {
        return encryption;
    }

    public String getPublicKey() {
        return publicKey;
    }

    public int getLocalPort() {
        return localPort;
    }

    public ConnectionMode getMode() {
        return mode;
    }

    public void send(IMessageComposer composer) {
        if (!composer.isComposed()) {
            composer.setComposed(true);
            composer.write();
        }

        try {
            channel.writeAndFlush(composer);
        } catch (Exception ignored) {
        }
    }

    public void initialiseEncryption() {
        if (encryption != null) {
            return;
        }

        encryption = new RC4();
        encryption.setKey(SecretKey.secretDecode(publicKey));

        if (Euclid.ENCRYPTION) {
            channel.pipeline().addFirst("encryption", new euclid.network.codec.EncryptionDecoder(encryption));
        }
    }

    public void onDisconnect() {
        if (disconnected) {
            return;
        }
        disconnected = true;
        player.onDisconnect();
    }

    public void disconnect() {
        if (disconnected) {
            return;
        }
        channel.close();
    }
}
